docs/e2e-testing-headscale-derper-tailscale.md

```diff
@@ -0,0 +1,368 @@
+# End to end testing: Tailscale, Headscale, Derper
+
+
+## Deploy environment
+
+For this testing, we'll use VMs on Azure.
+This ensures we can test scenarios that are close to real world,
+with public IP addresses and DNS names for Headscale and Derper.
+
+This document deploys everything in the `australiaeast` region;
+if you deploy in another region, substitute the name accordingly in domain names where referenced.
+
+### Terraform
+
+There is a terraform module at [./e2e.tf](./e2e.tf) that will deploy the environment.
+
+There are multiple ways to authenticate, but one way is to ensure the `az` command (azure cli) is installed and authenticated.
+See https://learn.microsoft.com/en-gb/cli/azure/get-started-with-azure-cli for more information.
+
+Then you will need `ARM_SUBSCRIPTION_ID` in the environment - see:
+https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/guides/azure_cli for details.
+Example:
+
+```bash
+export ARM_SUBSCRIPTION_ID=466ee356-ce1f-11ef-bf94-db2a042e145c
+```
+
+Then you can apply the terraform, providing required variables:
+
+```bash
+terraform apply -var "ssh_key=YOUR PUBLIC SSH KEY CONTENTS"
+```
+
+### Manually
+
+We'll deploy 5 VMs:
+
+| name (same DNS name) | full domain name                             | inbound rules                                                                                           | outbound rules |
+|----------------------|----------------------------------------------|---------------------------------------------------------------------------------------------------------|----------------|
+| headscale            | headscale.australiaeast.cloudapp.azure.com   | - TCP 22 (ssh)<br>- TCP 80 (letsencrypt challenge)<br>- TCP 443 (headscale)                             | None           |
```

Comment:
I don’t understand how these domain names are created and hosted. Are they only available in Azure, or are they accessible publicly?

Reply:
They're accessible publicly - Azure hosts the dns service. Headscale and Derper need domain names so we can reference them.